File: examples/python/casbot2_py_demo/casbot2_py_demo/action_voice_demo.py
import json

import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node

from crb_ros_msg.action import BasicActionPlay, VoicePlay
from crb_ros_msg.srv import ActionEvent, Voice


class ActionVoiceDemo(Node):
    def __init__(self) -> None:
        super().__init__("action_voice_demo")
        self.event_cli = self.create_client(ActionEvent, "/casbot/event_service")
        self.voice_cli = self.create_client(Voice, "/voice_svr")
        self.basic_action_cli = ActionClient(self, BasicActionPlay, "/basic_action_play")
        self.voice_action_cli = ActionClient(self, VoicePlay, "/action_voice_play")

    def call_service(self, cli, req, timeout_sec=3.0):
        if not cli.wait_for_service(timeout_sec=timeout_sec):
            self.get_logger().warn(f"服务不可用: {cli.srv_name}")
            return None
        fut = cli.call_async(req)
        rclpy.spin_until_future_complete(self, fut, timeout_sec=timeout_sec)
        return fut.result()

    def run(self) -> None:
        event_req = ActionEvent.Request()
        event_req.event_id = ""
        event_req.event_type = "ExecSkill"
        event_req.blocking = False
        event_req.param_json = json.dumps(
            {
                "payload": json.dumps({"action_type": "wave_hand"}),
                "target_tree": "basic_action_play",
            }
        )
        event_resp = self.call_service(self.event_cli, event_req)
        if event_resp:
            self.get_logger().info(
                f"event_service: success={event_resp.success}, msg={event_resp.message}"
            )

        voice_req = Voice.Request()
        voice_req.type = "question"
        voice_req.content_type = "text"
        voice_req.content = "你好"
        voice_resp = self.call_service(self.voice_cli, voice_req)
        if voice_resp:
            self.get_logger().info(f"voice_svr: success={voice_resp.success}")

        if self.basic_action_cli.wait_for_server(timeout_sec=2.0):
            goal = BasicActionPlay.Goal()
            goal.type = "wave_hand"
            fut = self.basic_action_cli.send_goal_async(goal)
            rclpy.spin_until_future_complete(self, fut, timeout_sec=3.0)
            self.get_logger().info("basic_action_play goal sent")
        else:
            self.get_logger().warn("basic_action_play action 不可用")

        if self.voice_action_cli.wait_for_server(timeout_sec=2.0):
            goal = VoicePlay.Goal()
            goal.wav_path = "test.wav"
            fut = self.voice_action_cli.send_goal_async(goal)
            rclpy.spin_until_future_complete(self, fut, timeout_sec=3.0)
            self.get_logger().info("voice_play goal sent")
        else:
            self.get_logger().warn("action_voice_play action 不可用")


def main() -> None:
    rclpy.init()
    node = ActionVoiceDemo()
    node.run()
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: examples/cpp/casbot2_cpp_demo/src/action_voice_demo.cpp
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "crb_ros_msg/action/basic_action_play.hpp"
#include "crb_ros_msg/action/voice_play.hpp"
#include "crb_ros_msg/srv/action_event.hpp"
#include "crb_ros_msg/srv/voice.hpp"

using BasicActionPlay = crb_ros_msg::action::BasicActionPlay;
using VoicePlay = crb_ros_msg::action::VoicePlay;
using namespace std::chrono_literals;

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("action_voice_demo");

  auto event_cli = node->create_client<crb_ros_msg::srv::ActionEvent>("/casbot/event_service");
  auto voice_cli = node->create_client<crb_ros_msg::srv::Voice>("/voice_svr");
  auto basic_action_cli = rclcpp_action::create_client<BasicActionPlay>(node, "/basic_action_play");
  auto voice_action_cli = rclcpp_action::create_client<VoicePlay>(node, "/action_voice_play");

  if (event_cli->wait_for_service(2s)) {
    auto req = std::make_shared<crb_ros_msg::srv::ActionEvent::Request>();
    req->event_id = "";
    req->event_type = "ExecSkill";
    req->blocking = false;
    req->param_json = "{\"payload\":\"{\\\"action_type\\\":\\\"wave_hand\\\"}\",\"target_tree\":\"basic_action_play\"}";
    auto fut = event_cli->async_send_request(req);
    if (rclcpp::spin_until_future_complete(node, fut, 3s) == rclcpp::FutureReturnCode::SUCCESS) {
      RCLCPP_INFO(node->get_logger(), "event_service: success=%s msg=%s",
        fut.get()->success ? "true" : "false", fut.get()->message.c_str());
    }
  }

  if (voice_cli->wait_for_service(2s)) {
    auto req = std::make_shared<crb_ros_msg::srv::Voice::Request>();
    req->type = "question";
    req->content_type = "text";
    req->content = "你好";
    auto fut = voice_cli->async_send_request(req);
    rclcpp::spin_until_future_complete(node, fut, 3s);
    if (fut.valid() && fut.wait_for(0s) == std::future_status::ready) {
      RCLCPP_INFO(node->get_logger(), "voice_svr: success=%s", fut.get()->success ? "true" : "false");
    }
  }

  if (basic_action_cli->wait_for_action_server(2s)) {
    BasicActionPlay::Goal goal;
    goal.type = "wave_hand";
    auto send_future = basic_action_cli->async_send_goal(goal);
    rclcpp::spin_until_future_complete(node, send_future, 3s);
    RCLCPP_INFO(node->get_logger(), "basic_action_play goal sent");
  }

  if (voice_action_cli->wait_for_action_server(2s)) {
    VoicePlay::Goal goal;
    goal.wav_path = "test.wav";
    auto send_future = voice_action_cli->async_send_goal(goal);
    rclcpp::spin_until_future_complete(node, send_future, 3s);
    RCLCPP_INFO(node->get_logger(), "voice_play goal sent");
  }

  rclcpp::shutdown();
  return 0;
}
